decidim/decidim
# Cast vote from the Voting Booth

#### :tophat: What? Why?
This enhances the Voting Booth to use the `decidim-bulletin_board` client to encrypt and cast the vote.
This is done in two different steps:

1. After the user answers all the questions we use the Bulletin Board JS client to encrypt the vote. Then we compute a `SHA-256` of the encrypted vote and sent both things to `decidim`. **Note:** The hash is computed in `decidim` code, but we are planning to move this computation to the JS client as well.
1. In the `decidim` server side we receive both the `encrypted_vote` and the `encrypted_vote_hash`. After checking if there is a match we store in the `votes` database table **only** the `encrypted_vote_hash`. Then we send the `encrypted_vote` to the Bulletin Board using the Ruby client.

With the `encrypted_vote_hash` the user will be able to check if the vote has been cast successfully in the Bulletin Board but this is something that will be addressed in a future PR.

#### :pushpin: Related Issues
- Related to https://github.com/decidim/decidim/issues/6625

#### Testing
* Setup an election and complete its key ceremony process.
* Wait until the election start_time.
* Enter to the voting booth and answer the election questions.
* Confirm the vote.
* Wait until the vote is encrypted and sent to the Bulletin Board.
* Use the [Bulletin Board GraphiQL interface](http://staging-bb-decidim.codegram.com/api/graphiql) to check that the vote was correctly stored using the following query (replace `[election_id]` with the id of the election present in the voting booth URL and `[hash]` with the hash given by the voting booth):

```graphql
query {
  logEntry(electionUniqueId: "decidim-test-authority.[election_id]", contentHash:"[hash]") {
    messageId
  }
}
```

#### :clipboard: Checklist
:rotating_light: Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.


### :camera: Screenshots
*Please add screenshots of the changes you're proposing*
![video](https://user-images.githubusercontent.com/106021/101799112-0d141b80-3b0c-11eb-8115-51fb09fe4612.gif)

:hearts: Thank you!
